sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1445,45 +1494,108 @@ _pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
                     : queue_type::LinkCopy;
 
   *QueueGroupOrdinal = Queue->Device->QueueGroup[QueueType].ZeOrdinal;
-
-  ze_command_queue_handle_t &ZeQueue = ZeQueues[CurrentIndex];
-  if (ZeQueue)
-    return ZeQueue;
-
-  // Command queue is not available at chosen index. So we create it below.
-  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
-
-  // Adjust the index to the L0 queue group since we represent "main" and "link"
+  // Adjust the index to the L0 queue group since we represent "main" and
+  // "link"
   // L0 groups with a single copy group ("main" would take "0" index).
   auto ZeCommandQueueIndex = CurrentIndex;
   if (QueueType == queue_type::LinkCopy && Queue->Device->hasMainCopyEngine()) {
     ZeCommandQueueIndex -= 1;
   }
+  *QueueIndex = ZeCommandQueueIndex;
 
-  ZeCommandQueueDesc.ordinal = *QueueGroupOrdinal;
-  ZeCommandQueueDesc.index = ZeCommandQueueIndex;
+  return CurrentIndex;
+}
+
+// Return a queue descriptor using the provided ordinal and index.
+void _pi_queue::pi_queue_group_t::createQueueDesc(
+    uint32_t Index, uint32_t Ordinal,
+    ZeStruct<ze_command_queue_desc_t> &ZeCommandQueueDesc) {
+
+  ZeCommandQueueDesc.ordinal = Ordinal;
+  ZeCommandQueueDesc.index = Index;
   ZeCommandQueueDesc.mode = ZE_COMMAND_QUEUE_MODE_ASYNCHRONOUS;
 
   // Evaluate performance of explicit usage for "0" index.
-  if (ZeCommandQueueIndex != 0) {
+  if (Index != 0) {
     ZeCommandQueueDesc.flags = ZE_COMMAND_QUEUE_FLAG_EXPLICIT_ONLY;
   }
 
   zePrint("[getZeQueue]: create queue ordinal = %d, index = %d "
           "(round robin in [%d, %d])\n",
           ZeCommandQueueDesc.ordinal, ZeCommandQueueDesc.index, LowerIndex,
           UpperIndex);
+}
+
+// This function will return one of possibly multiple available native
+// queues and the value of the queue group ordinal.
+ze_command_queue_handle_t &
+_pi_queue::pi_queue_group_t::getZeQueue(uint32_t *QueueGroupOrdinal) {
+
+  // QueueIndex is the proper L0 index.
+  // Index is the plugins concept of index, with main and link copy engines in
+  // one range.
+  uint32_t QueueIndex;
+  auto Index = getQueueIndex(QueueGroupOrdinal, &QueueIndex);
+
+  ze_command_queue_handle_t &ZeQueue = ZeQueues[Index];
+  if (ZeQueue)
+    return ZeQueue;
+
+  ZeStruct<ze_command_queue_desc_t> ZeCommandQueueDesc;
+  createQueueDesc(QueueIndex, *QueueGroupOrdinal, ZeCommandQueueDesc);
 
   auto ZeResult = ZE_CALL_NOCHECK(
       zeCommandQueueCreate, (Queue->Context->ZeContext, Queue->Device->ZeDevice,
                              &ZeCommandQueueDesc, &ZeQueue));
   if (ZeResult) {
     die("[L0] getZeQueue: failed to create queue");
   }
+
   return ZeQueue;
 }
 
+// This function will return one of possibly multiple available
+// immediate commandlists associated with this Queue.
+pi_command_list_ptr_t &
+_pi_queue::pi_queue_group_t::getImmCmdList(bool UseCopyEngine) {
```

Comment:
you shouldn't need the `UseCopyEngine` parameter here but instead check this group's `Type != queue_type::Compute`:

```suggestion
_pi_queue::pi_queue_group_t::getImmCmdList() {
```

Reply:
Yes